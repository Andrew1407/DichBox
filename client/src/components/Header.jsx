import React, { useContext, useCallback, useState, useEffect } from 'react';
import axios from 'axios';
import { motion, AnimatePresence } from 'framer-motion';
import { useHistory } from 'react-router-dom';
import { UserContext } from '../contexts/UserContext';
import { MenuContext } from '../contexts/MenuContext';
import { BoxesContext } from '../contexts/BoxesContext';
import { VerifiersContext } from '../contexts/VerifiersContext';
import { bouncingMotion, searchIconMotion } from '../styles/motions/bouncing-icons'
import { bounceDownMotion } from '../styles/motions/switch-icons';
import { headerMotion } from '../styles/motions/menu-components';
import defaultLogo from '../styles/imgs/default-user-logo.png';
import searchLogo from '../styles/imgs/search.png';
import showArrow from '../styles/imgs/menu-show.png';
import '../styles/header.css';

const Header = () => {
  const history = useHistory();
  const {
    username,
    setPathName,
    pathName,
    dispatchUserData
  } = useContext(UserContext);
  const {
    menuVisible,
    setMenuVisible,
    setSearchStr,
    searchStr,
    setUsersList,
    setFoundErr,
    foundErr,
    setLoading
  } = useContext(MenuContext);
  const { dispatchDataInput, cleanWarnings } = useContext(VerifiersContext);
  const { setBoxesList, setBoxHiddenState, setBoxDetails } = useContext(BoxesContext);
  const [searchInput, setSearchInput] = useState('');
  const [hidden, setHidden] = useState(false);

  const handleSearchClick = async () => {
    const inputValid = /\S+/.test(searchInput);
    if (!inputValid || foundErr) return;
    if (foundErr) setFoundErr(null);
    setSearchStr(searchInput);
    const searchBody = { searchStr: searchInput };
    setLoading(true);
    const { data } = await axios.post(`${process.env.APP_ADDR}/users/search`, searchBody);
    setLoading(false);
    const { searched } = data;
    if (searched) setUsersList(searched);
  };

  const handleSearchInput = e => {
    e.preventDefault();
    const input = e.target.value;
    setSearchInput(input);
  };

  const handleMenuClickClb = () => {
    if (foundErr) return;
    const currentUsername = username || '';
    const currentPathName = pathName || '';
    if (foundErr) setFoundErr(null);
    if (searchInput) setSearchInput('');
    if (searchStr) {
      setUsersList(null);
      setSearchStr('');
      return;
    }
    if (currentPathName === currentUsername) {
      setMenuVisible(!menuVisible);
    } else {
      setMenuVisible(true);
      setPathName(username);
      dispatchUserData({ type: 'CLEAN_DATA' });
      setBoxDetails({});
      setBoxesList([]);
      setBoxHiddenState(false);
      history.push('/');
    }
    dispatchDataInput({ type: 'CLEAN_DATA' });
    cleanWarnings();
  };
  const handleMenuClick = useCallback(
    handleMenuClickClb,
    [username, menuVisible, pathName, searchStr, searchInput, foundErr]
  );
  
  const backgroundColor = username ? 'rgb(50, 211, 240)' : 'grey';       //for image state

  useEffect(() => {
    if (searchStr && searchInput)
      handleSearchClick();
    else if (!searchInput)
      setSearchStr(null);
  }, [searchInput]);

  return (
    <div>
      <AnimatePresence>
      { !hidden && 
        <motion.div { ...headerMotion } id="header">
          <div  onClick={ () => setHidden(true) } id="header-name" className="highlight">
            <h1>DichBox</h1>
          </div>
          <motion.div { ...bouncingMotion } id="header-menu">
            <img src={ defaultLogo } className="highlight" onClick={ handleMenuClick } style={{ backgroundColor }} /> 
          </motion.div>
          <form id="header-search" className="highlight" onSubmit={ e => (e.preventDefault(), handleSearchClick()) }>
            <input value={ searchInput } type="text" placeholder="search users" onChange={ handleSearchInput } />
            <motion.img
              { ...searchIconMotion }
              src={ searchLogo } onClick={ handleSearchClick }
            />
          </form>
        </motion.div>
      }
      </AnimatePresence>
      { hidden &&
        <motion.img
          { ...bounceDownMotion }
          id="header-hidden" src={ showArrow } onClick={ () => setHidden(false) }
        />
      }
    </div>
  );
};

export default Header;
