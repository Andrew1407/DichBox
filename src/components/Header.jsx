import React, { useContext } from 'react';
import { MainContext } from '../contexts/MainContext';
import { useHistory } from 'react-router-dom';
import defaultLogo from '../styles/imgs/default-user-logo.png';
import '../styles/header.css';

const Header = () => {
  const { menuVisible, setMenuVisible, username, id } = useContext(MainContext);
  const history = useHistory();
  const handleMenuClick = e => {
    e.preventDefault()
    const currentPath = history.location.pathname;
    const pathName = new RegExp(username ? `^/${username}$` : '^/$');
    if (pathName.test(currentPath)) {
      setMenuVisible(!menuVisible);
    }
    else {
      setMenuVisible(true);
      history.push('/');
    }
  };
  const backgroundColor = id ? 'rgb(50, 211, 240)' : 'grey';       //for image state

  return (
    <div id="header">
      <div id="header-name">
        <h1 onClick={e => {e.preventDefault(); localStorage.setItem('username', 7890  )}} >DichBox</h1>
      </div>
      <div id="header-menu">
        <img src={defaultLogo}  onClick={ handleMenuClick } style={{ backgroundColor }} />
      </div>
      <div id="header-search">
        <input type="text" placeholder="search boxes" />
      </div>
    </div>
  )
};

export default Header;
