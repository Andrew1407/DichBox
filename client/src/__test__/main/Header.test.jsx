import React from 'react';
import { Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import Header from '../../components/Header';
import MenuContextProvider from '../../contexts/MenuContext';
import UserContextProvider from '../../contexts/UserContext';
import VerifiersContextProvider from '../../contexts/VerifiersContext';
import BoxesContextProvider from '../../contexts/BoxesContext';
import defaultLogo from '../../styles/imgs/default-user-logo.png';
import searchLogo from '../../styles/imgs/search.png';

describe('Header tests', () => {
  const { getByTestId, getByPlaceholderText } = render(
    <Router history={{ push: jest.fn(), location: { pathname: '' }, listen: jest.fn() }}>
      <MenuContextProvider>
      <UserContextProvider>
      <VerifiersContextProvider>
      <BoxesContextProvider>
        <Header />
      </BoxesContextProvider>
      </VerifiersContextProvider>
      </UserContextProvider>
      </MenuContextProvider>
    </Router>
  );
  const foundHeader = getByTestId('header-test');
  const foundSearchInput = getByPlaceholderText('search users');

  const tests = {
    'has title': () => {
      expect(foundHeader).toHaveTextContent('DichBox');
    },
    'has user button': () => {
      const butnImg = foundHeader.querySelector('#header-menu .highlight');
      expect(butnImg).toHaveAttribute('src', defaultLogo);
    },
    'has search area': () => {
      expect(foundSearchInput).toBeDefined();
      const [ serachImg ] = foundHeader.getElementsByTagName('img');
      expect(serachImg).toHaveAttribute('src', searchLogo);
    }
  };

  for (const key in tests)
    it(key, tests[key]);
});
