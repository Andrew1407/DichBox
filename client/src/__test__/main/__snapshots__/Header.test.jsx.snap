// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests matches snapshot 1`] = `
<div
  data-testid="header-test"
>
  <div
    id="header"
    style={
      Object {
        "transform": "translateY(-150px) translateZ(0)",
      }
    }
  >
    <div
      className="highlight"
      id="header-name"
      onClick={[Function]}
    >
      <h1>
        DichBox
      </h1>
    </div>
    <div
      id="header-menu"
      style={Object {}}
    >
      <img
        className="highlight"
        onClick={[Function]}
        src="test-file-stub"
        style={
          Object {
            "backgroundColor": "grey",
          }
        }
      />
    </div>
    <form
      className="highlight"
      id="header-search"
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="search users"
        type="text"
        value=""
      />
      <img
        onClick={[Function]}
        src="test-file-stub"
        style={
          Object {
            "transform": "skewX(25deg) translateZ(0)",
          }
        }
      />
    </form>
  </div>
</div>
`;
